proposals/required-members.md

```diff
@@ -0,0 +1,350 @@
+
+This is a simple hierarchy with 3 properties and 1 level of inheritance, but many real-world examples of these types of hierarchies go many levels deeper, accumulating larger and
+larger numbers of properties to pass along as they do so. Roslyn is one such codebase, for example, in the various tree types that make our CSTs and ASTs. This nesting is tedious
+enough that we have code generators to generate the constructors and definitions of these types, and many customers take similar approaches to the problem. C# 9 introduces records,
+which for some scenarios can make this better:
+
+```cs
+record Person(string FirstName, string LastName, string MiddleName = "");
+record Student(int ID, string FirstName, string LastName, string MiddleName = "") : Person(FirstName, LastName, MiddleName);
+```
+
+`record`s eliminate the first source of duplication, but the second source of duplication remains unchanged: unfortunately, this is the source of duplication that grows as the
+hierarchy grows, and is the most painful part of the duplication to fix up after making a change in the hierarchy as it required chasing the hierarchy through all of its locations,
+possibly even across projects and potentially breaking consumers.
+
+As a workaround to avoid this duplication, we have long seen consumers embracing object initializers as a way of avoiding writing constructors. Prior to C# 9, however, this had 2
+major downsides:
+
+1. The object hierarchy has to be fully mutable, with `set` accessors on every property.
+2. There is no way to ensure that every instantation of an object from the graph sets every member.
+
+C# 9 again addressed the first issue here, by introducing the `init` accessor: with it, these properties can be set on object creation/initialization, but not subsequently. However,
+we again still have the second issue: properties in C# have been optional since C# 1.0. Nullable reference types, introduced in C# 8.0, addressed part of this issue: if a constructor
+does not initialize a non-nullable reference-type property, then the user is warned about it. However, this doesn't solve the problem: the user here wants to not repeat large parts
+of their type in the constructor, they want to pass the _requirement_ to set properties on to their consumers. It also doesn't provide any warnings about `ID` from `Student`, as that
+is a value type. These scenarios are extremely common in database model ORMs, such as EF Core, which need to have a public parameterless constructor but then drive nullability of the
+rows based on the nullability of the properties.
+
+This proposal seeks to address these concerns by introducing a new feature to C#: required members. Required members will be required to be initialized by consumers, rather than by
+the type author, with various customizations to allow flexibility for multiple constructors and other scenarios.
+
+## Detailed Design
```

Comment:
It feels like a property without a `set` or `init` accessor can't be marked `required`. It might be good to specify this here.

Reply:
I actually don't agree. This is one of those "silly" scenarios, but it would mean that the user _must_ initialize it in their constructor.

Reply:
Though, perhaps we can say that a get-only non-auto-prop cannot be marked required.

Reply:
do we have a notion of an "assignable" property? that would be a property where either:

1. it has a `set` or `init` accessor,
2. it is a getter-only auto property, or
3. it is ref-returning

I suppose the most straightforward thing would be to allow all 3

Reply:
I'm not certain we have such a notion.